README.POD
=pod

README for DTA::CAB

=head1 ABSTRACT

DTA::CAB - "Cascaded Analysis Broker" for error-tolerant morphological analysis

=head1 REQUIREMENTS

=over 4

=item Gfsm

=item Encode

=item Storable

=item Tie::Cache

=item Getopt::Long

=item Pod::Usage

=item Time::HiRes

=item (a bunch of Gfsm transducers)

=back


=head1 DESCRIPTION

The DTA::CAB package provides an object-oriented compiler/interpreter for
error-tolerant heuristic morphological analysis of tokenized text.

=head1 SEE ALSO

=over 4

=item Gfsm(3perl)

URL: http://www.ling.uni-potsdam.de/~moocow/projects/gfsm

=item Gfsm::XL(3perl)

URL: http://www.ling.uni-potsdam.de/~moocow/projects/gfsm#gfsmxl

=back

=head1 AUTHOR

Bryan Jurish E<lt>jurish@bbaw.deE<gt>

=cut


Changes
##-*- Mode: Change-Log; coding: utf-8; -*-
##
## Change log for perl distribution DTA::CAB

UNRELEASED
Wed, 03 Mar 2021 14:09:32 +0100 moocow
	* Morph::Helsinki::DE updated tag-extraction heuristics (use original FST label conventions <+NN>, <+V> etc.)
	* added Morph::SMOR (just an alias for Morph::Helsinki::DE), Morph::TAGH (just an alias for Morph)

v1.114 Mon, 01 Mar 2021 15:45:44 +0100 moocow
	* added DTA::CAB::Chain::DE_free
	  - for use with (modified) Helsinki hfst-german FST from https://sourceforge.net/projects/hfst/files/resources/morphological-transducers
	* refactored DTA::CAB::Morph::Helsinki -> DTA::CAB::Morph::Helsinki::EN + DTA::CAB::Morph::Helsinki::DE

v1.113 Thu, 16 Jul 2020 15:42:53 +0200 moocow
	* added first-stab at DTA::CAB::Format:CONLLU conforming to https://universaldependencies.org/format.html
	  - optional special handling for some 'MISC' fields, including "json=JSON" for embedded TJ-format
	  - mostly based on TJ-format, with some "special fields"
	  - UNTESTED: development interrupted by svn server crash

v1.112 Mon, 30 Mar 2020 13:49:13 +0200 moocow
	* added %DTA::CAB::Analyzer::CLOSURE_CACHE - avoid re-compiling accessClosure() subs
	  - appears to fix memory-leak symptoms for Analyzer::Moot on debian stretch (perl 5.24.1)

v1.111 Thu, 23 Jan 2020 10:15:17 +0100 moocow
	* fixed normalized-key bug in DTA:CAB::Analyzer::DmootSub, ::MootSub
	  - earlier versions used normalized text as type-key ($dmoot->{tag},$moot->{word})
	  - old code caused bogus data-sharing for distinct input types mapped to same normalized form
 	  - bug observed for type-wise analysis of dwb1 lemmata ("aasz","asz") both mapped to "As", both assigned lemma ("aasz")
	    if processed in the same document, otherwise "asz" gets assigned lemma "asz"

v1.110 Mon, 23 Sep 2019 15:57:09 +0200 moocow
	* added http+unix support to dta-cab-check.perl
	* added CAB::Server::HTTP::Handler::sanitizeCharset() : tweak UTF-8 charset args to browser-friendly "utf-8"

v1.109 Tue, 26 Mar 2019 13:52:49 +0100 moocow
	* added Analyzer::DTAClean 'cleanPublic' option
	  - implements 'clean' option for public web-service
	  - still prunes 'unsafe' attributes (morph) including rwsub and dmootsub
	  - interest expressed by Graz users (G. Vogeler), 2019-03-19
	* added Format::CorpusExplorerPlugin (aliases "ceplugin", "ceplug")
	* added Format::Raw::Base
	  - common base class for Format::Raw::*, supports simple "normalized plain text" output

v1.108 Thu, 21 Mar 2019 12:59:49 +0100 moocow
	* added ling2norm.xsl, ling2plain.xsl
	* added format aliases teiws-names, teiws-ling-names (input option {teinames=>1})

v1.107 Fri, 22 Feb 2019 09:40:00 +0100 moocow
	* added DTA::TokWrap and GermaNet::Flat dependencies (used by built-in analyzer classes)

v1.106 2019-02-12  moocow
	* improved Version.pm (re-)generation: only if this looks like a "proper" checkout
	* added Changes (this file: extracted from SVN logs & reformatted)
	* cleanup for CPAN release
	* SVNVERSION tweaks (revision only, no root URL)
	* find.hack: File::Find hacks for ExtUtils::Manifest
	* removed some (but not all) doubled and/or recursuive symlinks from SVN
	  - they don't play nicely with ExtUtils::Manifest / MakeMaker / File::Find

v1.105 2019-01-09  moocow
	* added ddc full lemma-list (LemmaListAll LemmasAll llist-all ll-all lla lemmas lemmata)

v1.104 2018-12-17  moocow
	* default -log-watch=USR1 for dta-cab-server.sh
	* added server logInitAnalyzer option
	* added -log-watch=SIGNAL syntax (reload log-config on user signal, e.g. -log-watch=USR1)

v1.103 2018-12-06  moocow
	* XmlLing : escape token text if not running in twcompat mode
	* syslog debugging
	* added cab-syslog.l4p -- getting weird rsyslog errors
	  > Oct 25 13:55:12 plato liblogging-stdlog: action 'action 0' resumed (module 'builtin:ompipe') [v8.24.0 try http://www.rsyslog.com/e/2359 ]
	  > Oct 25 13:55:51 plato liblogging-stdlog: action 'action 1' resumed (module 'builtin:ompipe') [v8.24.0 try http://www.rsyslog.com/e/2359 ]
	  ... on every message; not pretty
	* systemd-friendliness for cab sysv-scripts (control groups, etc)
	* dta-cab.sh: merged changes from bogus for in dstar/cabx/
	* added cgiwrap for version
	* web-howto typos
	* updated 'fliegen' example in web-howto
	* clean Version.pm
	* WebServiceHowto updates for XmlLing
	* alias tweaks
	* XmlLing for server mode
	* added support for TEI att.linguistic features
	  - new formatter Format::XmlLing (flat att.linguistic features, with optional TokWrap compatibility for later spliceback)
	  - new TEI and TEIws options 'att.linguistic=bool' : force use of XmlLing sub-formatter with appropriate options
	  - new TEI and TEIws aliases (ltei ... ling-tei-xml, lteiws ... ling-tei-ws)
	  - updated Format SUBCLASSES docs and examples
	  - still TODO: integrate new formats into CAB demo web-GUI and HOWTO
	* added format XmlLing: use TEI att.linguistic attributes

v1.102 2018-06-20  moocow
	* howto updates
	  for spliced2ling
	* added
	  spliced2ling xsl stuff
	* HttpProtocol.pod: added explicit
	  'xpost' reference
	* DSGVO stuff
	* clean Version.pm
	* attempt to ensure Listen=SOMAXCONN for
	  DTA::CAB::Server::HTTP::UNIX

v1.101 2018-04-13  moocow
	* dta-cab-server.sh: handle tcp<->unix relay via new variables
	  + added -verbose LEVEL option for debugging
	  + added 'config|debug' action to view configuration variables
	* system/xlit-unix.plm: test tcp relay handling by sysv-like dta-cab-server.sh
	* more cab-v1.101 check tweaks
	  (icinga/pnp4nagios doesn't like floats in engineering notation)
	* dta-cab-http-check.perl: v1.101 perfdata fixes
	* status.html.tpl: compatibility fixes for transition
	* added rss and exponential moving average query times to CAB status output
	  - implements mantis #26054

v1.100 2018-03-21  moocow
	* dta-cab-server.sh:
	  - disable watchdog by default (let icinga do this)
	  - use administrative lock-files to avoid concurrent operations
	* minor tempfile tweaks attempting to get at mantis #25739


demo.js
//-*- Mode: Javascript; coding: utf-8; c-basic-offset: 2; -*-

var uploadMode;

//======================================================================
// form data read/write

//--------------------------------------------------------------
//-- nod = getid(id)
function getid(id) {
    return document.getElementById(id);
}

//-- nod = getname(name)
function getname(name) {
    return document.getElementsByName(name)[0];
}

//--------------------------------------------------------------
// val = valGet(nod_or_id_or_name)
function valGet(id) {
    var nod = id;
    if (typeof(nod) == 'string') {
	nod = getid(id);
	if (nod==null) {
	    nod = getname(id);
	}
    }
    if (nod==null) {
	throw("valGet(): no node for id="+id);
	return null;
    }
    var ttag = nod.tagName.toLowerCase();

    if (ttag=="input" || ttag=="select" || ttag=="textarea") {
	var ttype = nod.getAttribute('type');
	if (ttype==null) { ttype=ttag; }
	ttype = ttype.toLowerCase();
	if (ttype=="checkbox")    { return nod.checked; }
	else if (ttype=="select") { return nod.options[nod.selectedIndex].value; }
	else { return nod.value; }
    }
    return nod.innerHTML;
}

//--------------------------------------------------------------
// undef = valSet(nod_or_id_or_name,val)
function valSet(id,val) {
    var nod = id;
    if (typeof(nod) == 'string') {
	nod = getid(id);
	if (nod==null) {
	    nod = getname(id);
	}
    }
    if (nod==null) {
	throw("valSet(): no node for id="+id+", val="+val);
	return null;
    }
    if (val==null) {
	val = "";
    }
    var ttag = nod.tagName.toLowerCase();
    if (ttag=="input" || ttag=="select") {
	var ttype = nod.getAttribute('type');
	if (ttype==null) { ttype = ttag; }
	ttype = ttype.toLowerCase();
	if (ttype=="checkbox")    { nod.checked = Boolean(val); }
	else if (ttype=="select") {
	    for (opti in nod.options) {
		if (nod.options[opti].value == val) {
		    nod.selectedIndex = opti;
		    break;
		}
	    }
	}
	else { nod.value = String(val); }
    } else {
	nod.innerHTML = String(val);
    }
}

//======================================================================
// generic XMLHttpRequest stuff

//-- encstr = formString({param:val,...})
function formString(params) {
    var form = [];
    for (var p in params) {
	if (params[p] == null) { continue; }
	form.push(p+'='+encodeURIComponent(params[p]));
    }
    return form.join('&');
}

//-- uristr = getURI(base,{param:val,...})
function getURI(base,params) {
    return base + '?' + formString(params);
}

//-- undef = httpGet(uri,onReadyCallback)
//  + onReadyCallback = function(req) { ... }
function httpGet(uri,onReadyCallback) {
    var get_uri = uri;
    var req = new XMLHttpRequest();
    req.open("GET",get_uri,true);
    req.onreadystatechange=function(){onReadyCallback(req)};
    req.send(null);
}

//======================================================================
// cab request (live query or upload)

var cab_url_base = '/query';

function cabQuery() {
    //-- common parameters
    var params = {
	"a" : valGet('a'),
	"fmt" : valGet('fmt'),
	"clean" : Number(valGet('clean')),
    };
    //-- flags
    if (valGet('pretty')) {
	params['pretty'] = "1";
	params['raw']    = "1";
    }
    if (!valGet('exlex_enabled')) {
	params['exlex_enabled']  = "0";
	params['static_enabled'] = "0";
    }

    //-- options
    try {
	var opts_s = valGet('qo');
	var opts = JSON.parse(opts_s);
	if (opts != null) {
	    for (var o in opts) {
		if (opts[o] == null) {
		    delete params[o];
		} else {
		    params[o] = opts[o];
		}
	    }
	}
    }
    catch (err) {
	clog("cabQuery(): error parsing user options: " + err.description);
    }

    //-- query or upload
    if (Boolean(uploadMode)) {
	return cabUpload(params);
    } 

    //-- query mode
    var q = valGet('q')
    if (valGet('tokenize')) {
	params['q'] = q;
    } else {
	params['qd'] = q;
    }

    //-- guts
    var caburi  = getURI(cab_url_base,params);
    var cablink = getid('cabLink');
    cablink.href      = caburi;
    cablink.innerHTML = caburi;

    valSet('cabData', 'Querying ' + caburi);
    httpGet(caburi,
	    function(req){ if (req.readyState==4) { valSet('cabData',req.responseText); } });
    
    //-- traffic-light query
    cabTrafficQuery(q, 'btn_trf_cabq');

    return false; //-- return false to disable "real" form submission
}

//======================================================================
// cab file upload (called by cabQuery)

function cabUpload(params) {
    //-- common vars
    var qfnod = getid('i_qf');

    //-- sanity check(s) 
    if (!Boolean(qfnod.value)) {
	alert("No file selected for upload!");
	return false; //-- return false to disable "real" form submission
    }

    //-- tweaks: auto-format (use file extension)
    var fmt     = params['fmt'];
    var autofmt = (fmt == 'auto');
    if (autofmt) {
	fmt = qfnod.value.match(/[^\.]*$/)[0].toLowerCase();
    }

    //-- tweaks: tokenization
    var istei = (fmt.search(/tei/) != -1);
    var dotok = valGet('tokenize');

    qfnod.name = 'qd';
    if (!dotok && istei) {
	fmt = 'teiws';
    }
    else if (dotok && !istei && (!autofmt || (fmt=='txt'))) {
	qfnod.name = 'q';
    }

    //-- construct base URI
    delete params['q'];
    delete params['qd'];
    params['fmt'] = fmt;
    params['qname'] = qfnod.name;
    params['qfile'] = qfnod.value;
    var caburi = getURI(cab_url_base,params).replace(/&/g, ';');

    //-- tweak query form
    var qform = getid('queryForm');
    qform.action = caburi;
    qform.method = 'POST';
    qform.enctype = 'multipart/form-data';

    // Strangeness Fri, 17 May 2013 14:13:19 +0200:
    //  - returning true here causes weird 'Client closed' errors from CAB server on post-submit 'Back' clicks from chromium
    //  - iceweasel/firefox seem unaffected
    return true;     //-- true: enable "real" form submission
}

function qformEnable(enabled) {
    var disabled = !Boolean(enabled);
    getname('a').disabled = disabled;
    getname('fmt').disabled = disabled;
    getname('pretty').disabled = disabled;
    getname('clean').disabled = disabled;
    getname('exlex_enabled').disabled = disabled;
    getname('tokenize').disabled = disabled;
    getname('qo').disabled = disabled;
    getname('_s').disabled = disabled;
}

//======================================================================
// CAB auto-query ("traffic lights")

function cabTrafficQuery(w,spanid) {
    var nod = getid(spanid);
    nod.style.backgroundColor = '#c0c0c0'; //-- unknown: white
    nod.title = nod.title.replace(/safe|unsafe|unrecognized|unknown/g, 'unknown');
    if (w.search(/\s/) != -1) { return; } //-- word with spaces: ignore it
    httpGet(getURI(cab_url_base,
		   {"qd":JSON.stringify({"text":w}),
		    "fmt":"json",
		    "a":"default.msafe",
		    "exlex_enabled":0,
		    "clean":0,
		   }),
	    function(req){ cabTrafficReady(w,spanid,req) });
}

var cab_autobg_red    = '#ff0000';
var cab_autobg_yellow = '#ffff00';
var cab_autobg_green  = '#00ff00';
function cabTrafficReady (w,spanid,req) {
    if (req.readyState!=4) { return; }
    if (req.status>=400) {
	clog('cabAutoReady('+w+'): HTTP Error: '+req.status);
	return;
    }
    var rsp = JSON.parse(req.responseText);
    var tok = rsp.body[0].tokens[0];
    var nod = getid(spanid);
    var xlit_id    = (tok.text == tok.xlit.latin1Text);
    var has_morph  = (tok.hasmorph || (tok.morph!=null && tok.morph.length>0));
    var has_mlatin = (tok.mlatin!=null && tok.mlatin.length>0);
    if (w.search(/\s/) != -1) {
	; //-- do nothing
    }
    else if (xlit_id && Boolean(tok.msafe) && has_morph) {
	nod.style.backgroundColor = cab_autobg_green;
	nod.title = nod.title.replace(/safe|unsafe|unrecognized|unknown/g, 'safe');
    }
    else if (xlit_id && (has_morph || has_mlatin)) {
	nod.style.backgroundColor = cab_autobg_yellow;
	nod.title = nod.title.replace(/safe|unsafe|unrecognized|unknown/g, 'unsafe');
    }
    else {
	nod.style.backgroundColor = cab_autobg_red;
	nod.title = nod.title.replace(/safe|unsafe|unrecognized|unknown/g, 'unrecognized');
    }
}

//======================================================================
// debugging

function clog(msg) {
    if (typeof(console)!='undefined' && console!=null) {
	//-- javascript log
	console.log(msg);
    }
    //-- local log
    var lnod = document.getElementById('logData');
    if (lnod != null) {
	valSet(lnod,valGet(lnod)+msg+'\n');
    }
}

//======================================================================
// query parsing

// map = parseQuery(queryStr)
function parseQuery(qs) {
    var qm = {};
    if (qs.substr(0,1)=='?') { qs=qs.substr(1); } //-- strip leading "?"
    var ql = qs.split("&");
    for (var i=0; i<ql.length; i++) {
	if (ql[i] == "") { continue; }
	var kv = ql[i].split("="); 
	qm[decodeURIComponent(kv[0])] = decodeURIComponent(kv[1].replace(/\+/g,' '));
    }
    return qm;
}


//======================================================================
// CAB demo stuff

function cabDemoInit() {
    var nod = getid('cabLink');
    cab_url_base = 'http://' + window.location.host + '/query';
    nod.href = cab_url_base;
    nod.innerHTML = cab_url_base;

    //-- initialize
    var qm = parseQuery(window.location.search);
    var qmkeys = ["q","a","fmt","pretty","clean","exlex_enabled","tokenize","qo"];
    for (var i=0; i<qmkeys.length; i++) {
	if (qm[qmkeys[i]] != null) { valSet(qmkeys[i], qm[qmkeys[i]]); }
    }
    if (String(valGet('q')) != "") { cabQuery(); }

    return;
}
